### PR TITLE
Add the Response-Behavior on PATCH to decide which kind of results to get back

### PR DESCRIPTION
**Pull request by Natim**
r? @ametaireau

### CODE REVIEW

File: docs/resource.rst [hunk 1]
```diff
@@ -306,7 +307,146 @@ of the API, the following fields are also modifiable:
 - ``resolved_url``
 - ``resolved_title``
 
-**Errors**
+Response behavior
+-----------------
+
+When doing a PATCH you can choose which kind of response you want to get.
+
+You have got 3 behaviors:
```
Comment (ametaireau) [inline docs/resource.rst]: 3 behaviors are defined

File: docs/resource.rst [hunk 2]
```diff
@@ -306,7 +307,146 @@ of the API, the following fields are also modifiable:
 - ``resolved_url``
 - ``resolved_title``
 
-**Errors**
+Response behavior
+-----------------
+
+When doing a PATCH you can choose which kind of response you want to get.
+
```
Comment (leplatrem) [inline docs/resource.rst]: Who are you?

File: readinglist/tests/test_views_article.py [hunk 3]
```diff
@@ -168,6 +173,26 @@ def test_timestamp_is_not_updated_if_already_unread(self):
...
                          self.record['last_modified'])
 
+    def test_body_behavior_set_to_diff_return_only_diff(self):
+        body = {'unread': "True", 'read_position': 10}
+        self.headers.update({
+            'Response-Behavior': 'diff'
+        })
+        resp = self.app.patch_json(self.url,
+                                   body,
+                                   headers=self.headers)
```
Comment (leplatrem) [inline readinglist/tests/test_views_article.py]: Should not be 10?

Comment (Natim) [inline readinglist/tests/test_views_article.py]: When back to Unread the ``read_position`` is reset

File: docs/resource.rst [hunk 4]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
...
+    {
+        "marked_read_by": "Ipad", 
+        "marked_read_on": 1425316211577, 
+        "unread": false
+    }
+
+Calling with ``Response-Behavior: diff``
+::::::::::::::::::::::::::::::::::::::::
+
```
Comment (Natim) [inline docs/resource.rst]: Missing ``::``

File: readinglist/views/article.py [hunk 5]
```diff
@@ -129,3 +129,33 @@ def collection_post(self, *args, **kwargs):
...
+        else:
+            new_record = record
+
+        if body_behavior == 'full':
+            return new_record
+        elif body_behavior == 'light':
+            return {k: new_record[k] for k in changed}
+        elif body_behavior == 'diff':
+            return {k: new_record[k] for k in changed
+                    if native_value(changes.get(k)) != new_record.get(k)}
```
Comment (leplatrem) [inline readinglist/views/article.py]: Why not having overriden "update_record" instead ?

File: readinglist/tests/test_views_article.py [hunk 6]
```diff
@@ -168,6 +173,26 @@ def test_timestamp_is_not_updated_if_already_unread(self):
         self.assertEqual(resp.json['last_modified'],
                          self.record['last_modified'])
 
+    def test_body_behavior_set_to_diff_return_only_diff(self):
```
Comment (leplatrem) [inline readinglist/tests/test_views_article.py]: Add a "spec" test to assert that default is full

File: docs/resource.rst [hunk 7]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
 - ``resolved_url``
 - ``resolved_title``
 
-**Errors**
+Response behavior
+-----------------
+
+On a PATCH it is possible to customize which kind of response is wanted.
```
Comment (leplatrem) [inline docs/resource.rst]: to choose among different behaviors for the response content.

File: docs/resource.rst [hunk 8]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
...
+
+Three behaviors are available:
+
+- ``full``: The default behavior, give you back the full new record.
```
Comment (leplatrem) [inline docs/resource.rst]: Returns the whole record (**default**)

File: docs/resource.rst [hunk 9]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
...
+- ``light``: Give you back the new fields values modified by your request
```
Comment (leplatrem) [inline docs/resource.rst]: ``Returns only the fields whose value was changed``

File: docs/resource.rst [hunk 10]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
...
+- ``diff``: Give you back the new fields values that don't match your request fields values.
```
Comment (leplatrem) [inline docs/resource.rst]: ``Returns only the fields values that don't match those provided``

File: docs/resource.rst [hunk 11]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
...
+
+.. code-block:: http
+    POST /v0/articles HTTP/1.1
+    Accept: application/json
+    Accept-Encoding: gzip, deflate
+    Authorization: Basic TmF0aW06
+    Content-Length: 71
+    Content-Type: application/json; charset=utf-8
+    Host: localhost:8000
+    User-Agent: HTTPie/0.8.0
```
Comment (leplatrem) [inline docs/resource.rst]: replace all those lines by ``...``

File: docs/resource.rst [hunk 12]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
...
+
+Three behaviors are available:
+
+- ``full``: The default behavior, give you back the full new record.
+- ``light``: Give you back the new fields values modified by your request
+- ``diff``: Give you back the new fields values that don't match your request fields values.
+
+As an example:
+
+Starting with::
```
Comment (leplatrem) [inline docs/resource.rst]: For example, using the default behavior :

File: docs/resource.rst [hunk 13]
```diff
@@ -302,7 +303,146 @@ of the API, the following fields are also modifiable:
...
+        "status": 0, 
+        "stored_on": 1425315406801, 
+        "title": "Trunat", 
+        "unread": true, 
+        "url": "http://www.trunat.fr", 
+        "word_count": null
+    }
+
+Calling with ``Response-Behavior: light``
+:::::::::::::::::::::::::::::::::::::::::
```
Comment (leplatrem) [inline docs/resource.rst]: Using ``Response-Behavior: light``

**Comment by leplatrem** [comment]
r+

File: readinglist/views/article.py (hunk 5 above)
Comment (Natim) [inline readinglist/views/article.py]: How?

Comment (Natim) [inline readinglist/views/article.py]: Ok see https://github.com/mozilla-services/cliquet/pull/61

**Pull request merged**